src/Rules/PhpDoc/IncompatiblePhpDocInheritedReturnType.php

```diff
@@ -0,0 +1,114 @@
+<?php declare(strict_types=1);
+
+namespace PHPStan\Rules\PhpDoc;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MissingMethodFromReflectionException;
+use PHPStan\Reflection\Php\PhpMethodReflection;
+use PHPStan\Type\FileTypeMapper;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+
+class IncompatiblePhpDocInheritedReturnType implements \PHPStan\Rules\Rule
+{
+
+	/** @var FileTypeMapper */
+	private $fileTypeMapper;
+
+	public function __construct(FileTypeMapper $fileTypeMapper)
+	{
+		$this->fileTypeMapper = $fileTypeMapper;
+	}
+
+	public function getNodeType(): string
+	{
+		return \PhpParser\Node\FunctionLike::class;
+	}
+
+	/**
+	 * @param Node|Node\Stmt\ClassMethod $node
+	 * @param \PHPStan\Analyser\Scope    $scope
+	 *
+	 * @return string[]
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$parents = $scope->getClassReflection()->getParents();
+
+		if (!count($parents)) {
+			return [];
+		}
+
+		$docComment = $node->getDocComment();
+		if ($docComment === null) {
+			return [];
+		}
+
+		$resolvedPhpDoc = $this->fileTypeMapper->getResolvedPhpDoc(
+			$scope->getFile(),
+			$scope->getClassReflection()->getName(),
+			$docComment->getText()
+		);
+
+		try {
+			/** @var PhpMethodReflection $parentMethod */
+			$parentMethod = $parents[0]->getMethod($node->name, $scope);
+		} catch (MissingMethodFromReflectionException $e) {
+			return [];
+		}
+
+		$resolvedParentPhpDoc = $this->fileTypeMapper->getResolvedPhpDoc(
+			$scope->getFile(),
+			$parents[0]->getName(),
+			$parentMethod->getDocComment()
+		);
+
+		$errors = [];
+
+		if ($resolvedPhpDoc->getReturnTag() !== null) {
+			$phpDocReturnType = $resolvedPhpDoc->getReturnTag()->getType();
+			$phpDocParentReturnType = $resolvedParentPhpDoc->getReturnTag()->getType();
+
+			if (!($phpDocReturnType instanceof UnionType) && $phpDocReturnType !== $phpDocParentReturnType) {
```

Comment:
`Type` instances cannot be compared by identity. `\PHPStan\Type\Type::isSuperTypeOf` must be used instead.